File: queues/ltcd_2528.py
'''
LeetCode - problem 2528
'''

    def maxPower(self, stations: List[int], r: int, k: int) -> int:
        lo, hi = min(stations), sum(stations) + k
        ans = lo
        while lo <= hi:
            mid = (lo + hi) // 2
            ts, left = stations[:], k
            cur = sum(ts[:r + 1])
            if cur < mid:
                left -= mid - cur
                ts[r] += mid - cur
                cur = mid
            for i in range(1, len(ts)):
                if i + r < len(ts): cur += ts[i + r]
                if i - r - 1 >= 0: cur -= ts[i - r - 1]
                if cur < mid:
                    left -= mid - cur
                    ts[min(i + r, len(ts) - 1)] += mid - cur
                    cur = mid
                if left < 0: break
            if left >= 0:
                ans = max(ans, mid)
                lo = mid + 1
            else: hi = mid - 1
        return ans

File: queues/ltcd_2528.cpp
//LeetCode - problem 2528

    bool isPowerable(vector<int> stations, long long minPower, int k, int r) {
        int start = -1*r;
        int end = r;
        int n = stations.size();
        
        long long current_pow = 0;
        for(int i=0; i<=r && i<n; i++){
            current_pow += stations[i];
        }
        
        int mid = 0;
        while(mid<n) {
            if(current_pow < minPower) {
                long long diff = minPower - current_pow;
                if(diff>k)return false;
                k -= diff;
                stations[min(mid+r, n-1)] += diff;
            }
            current_pow = max(current_pow, minPower);
            current_pow -= start>=0?stations[start]:0;
            current_pow += end<=n-2?stations[end+1]:0;
            start++;
            end++;
            mid++;
        }
        
        return true;
    }
    long long maxPower(vector<int>& stations, int r, int k) {
     
        long long start = 0;
        long long end = LONG_LONG_MAX;
        long long maxPow = 0;
        
        while(start<=end) {
            long long mid = (start + end)/2;
            if(isPowerable(stations, mid, k, r)) {
                start = mid+1;
                maxPow = max(maxPow, mid);
            }else {
                end = mid-1;
            }
        }
        
        return maxPow;
        
    }
